### PR TITLE
Unsafe row

### PR DESCRIPTION
**Pull request by tpoterba**

### CODE REVIEW

**Comment by tpoterba** [comment]
Note that I am not at all convinced that the random tests are providing good coverage: https://github.com/hail-is/hail/issues/1894

**Comment by cseed** [comment]
Aww, yes!  Very exciting!  I will try to go through this today.

File: src/test/scala/is/hail/annotations/UnsafeSuite.scala [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+import is.hail.check.Arbitrary._
+import is.hail.expr._
+import org.apache.spark.sql.Row
+import org.testng.annotations.Test
+
+class UnsafeSuite extends SparkSuite {
+
+  val g = for {
+    t <- Type.genStruct
+    a <- t.genNonmissingValue} yield (t, a)
```
Comment (danking) [inline src/test/scala/is/hail/annotations/UnsafeSuite.scala]: We should give much less size to the type than the value. At the least, something like the following will ensure their combined size is O(size).

```scala
def genStructTypeAndNonMissingValue: Gen[(Type, Annotation)] = for {
  (x, y) <- Gen.squareOfAreaAtMostSize
  t <- Type.genStruct.resize(x)
  v <- t.genNonMissingValue(y)
} yield (t, v)
```

File: src/main/scala/is/hail/expr/Type.scala [hunk 2]
```diff
@@ -1184,4 +1232,24 @@ case class TStruct(fields: IndexedSeq[Field]) extends Type {
...
+        offset += (fAlignment - mod)
+      a(f.index) = offset
+      offset += fSize
+    }
+    a
+  }
+
+  override lazy val byteSize: Int = if (size == 0) 0 else byteOffsets.last + fields.last.typ.byteSize
+
+  override def alignment: Int = 8
```
Comment (tpoterba) [inline src/main/scala/is/hail/expr/Type.scala]: min 4, though (missing bytes are read as int)

Comment (cseed) [inline src/main/scala/is/hail/expr/Type.scala]: Ah!  Yes.  The missing bits are another field.  Should we store missing bits as bytes?  Only can I can see where it matters is small structs of Booleans.  Probably doesn't matter.

File: src/main/scala/is/hail/annotations/UnsafeAnnotations.scala [hunk 3]
```diff
@@ -0,0 +1,402 @@
...
+  def roundUpAlignment(offset: Int): Int = {
+    val mod = offset & 0x7
+    if (mod != 0)
+      offset + (8 - mod)
+    else
+      offset
+  }
+}
+
+class UnsafeRowBuilder(t: TStruct, sizeHint: Int = 0, debug: Boolean = false) {
```
Comment (cseed) [inline src/main/scala/is/hail/annotations/UnsafeAnnotations.scala]: I think adding a little additional abstraction will make all this cleaner.  I suggest:

```
// I'm not sure about alignment guarantees but Long seems safer
final class MemoryBlock(mem: Array[Long]) {
  def loadInt(off: Int): Int = ...
  // same for Byte, Short, Long, Float, Double, BitArray
  def reallocate(): MemoryBlock
}

final class Pointer(mem: MemoryBlock, memOffset: Int) {
  def loadInt(): Int = loadInt(0)
  def loadInt(off: Int): Int = mem.loadInt(off + memOffset)
  // and others.  No reallocate.

  def offset(off: Int): Pointer = new Pointer(mem, memOffset + off)
}

final class MemoryBuffer(mem: MemoryBlock, offset: Int) {
  // expose MemoryBlock operations, plus:
  def appendInt(i: Int) // reallocate if necessary
  // other types...
  def allocate(nBytes: Int): Int // reallocate if necessary, return offset
  def align(alignment: Int)
}
```

File: src/main/scala/is/hail/annotations/UnsafeAnnotations.scala [hunk 4]
```diff
@@ -0,0 +1,402 @@
...
+        dictKeyBuffer.clear()
+        dictValueBuffer.clear()
+        m.foreach { case (k, v) =>
+            dictKeyBuffer += k
+            dictValueBuffer += v
+        }
+        putArray(dictKeyBuffer, offset, kt)
+        putArray(dictValueBuffer, offset + 4, vt)
+      case struct: TStruct =>
+        if (struct.size > 0)
```
Comment (cseed) [inline src/main/scala/is/hail/annotations/UnsafeAnnotations.scala]: Just let putStruct handle this.

File: src/main/scala/is/hail/annotations/UnsafeAnnotations.scala [hunk 5]
```diff
@@ -0,0 +1,402 @@
...
+        putInt(newBits, intIndex)
+      } else {
+        val shift = put(elt, cursor, elementType)
+      }
+      cursor += eltSize
+      i += 1
+    }
+  }
+
+  private def putStruct(value: Row, offset: Int, struct: TStruct) {
```
Comment (cseed) [inline src/main/scala/is/hail/annotations/UnsafeAnnotations.scala]: I can't quite figure out how this code works.  I expected putStruct to allocate the space for the struct, and then put each field, with arrays and things being added after the struct.  You write each field, I just don't see the allocate.  Am I missing something?

**Comment by cseed** [comment]
@tpoterba Back to you.  Great first cut.  Address the comments and I think we'll be pretty close.

File: src/main/scala/is/hail/expr/Type.scala (hunk 2 above)
Comment (tpoterba) [inline src/main/scala/is/hail/expr/Type.scala]: eh, bytes is better. There's no reason not to.

File: src/main/scala/is/hail/annotations/UnsafeAnnotations.scala (hunk 3 above)
Comment (tpoterba) [inline src/main/scala/is/hail/annotations/UnsafeAnnotations.scala]: is the intent for one memoryBlock to store multiple records here?

Comment (cseed) [inline src/main/scala/is/hail/annotations/UnsafeAnnotations.scala]: Do you mean temporally or spatially?  I guess the answer is yes to both, first temporally, we should reuse MemoryBuffers for successive records, this is the same as the mutable stuff we started on, and then spatially we can decode and process multiple records at a time instaed of using iterators per-record (what database people call vectorization).

File: src/main/scala/is/hail/annotations/UnsafeRow.scala [hunk 6]
```diff
@@ -0,0 +1,145 @@
+package is.hail.annotations
+
+import is.hail.expr._
+import is.hail.variant.{AltAllele, Genotype, Locus, Variant}
+import org.apache.spark.sql.Row
+
+class UnsafeRow(@transient var t: TStruct, val mb: MemoryBlock, val mbOffset: Int, debug: Boolean = false) extends Row {
```
Comment (cseed) [inline src/main/scala/is/hail/annotations/UnsafeRow.scala]: I don't think we should leave in `debug`.  Thoughts?

File: src/main/scala/is/hail/annotations/UnsafeRowBuilder.scala [hunk 7]
```diff
@@ -0,0 +1,359 @@
...
+      case struct: TStruct =>
+        putStruct(value.asInstanceOf[Row], offset, struct)
+      case TAltAllele | TVariant | TGenotype | TLocus | TInterval =>
+        val expandedType = Annotation.expandType(elementType).asInstanceOf[TStruct]
+        if (debug)
+          println(s"putting expanded struct $expandedType at $offset")
+        val expandedAnnotation = Annotation.expandAnnotation(value, elementType).asInstanceOf[Row]
+        putStruct(expandedAnnotation, offset, expandedType)
+
+      case err => throw new NotImplementedError(err.toPrettyString(compact = true))
```
Comment (cseed) [inline src/main/scala/is/hail/annotations/UnsafeRowBuilder.scala]: Match failure seems like a fine error here.

File: src/main/scala/is/hail/annotations/UnsafeRowBuilder.scala [hunk 8]
```diff
@@ -0,0 +1,359 @@
...
+
+  private def setMissing(index: Int) {
+    val byteIndex = index >> 3
+    val bitIndex = index & 0x7
+
+    val oldBits = buffer.loadByte(byteIndex)
+    buffer.storeByte(byteIndex, (oldBits | (0x1 << bitIndex)).toByte)
+  }
+
+//  private def copyStructFromPointer(st: TStruct, o1: Int, o2: Int, p: Pointer) {
```
Comment (cseed) [inline src/main/scala/is/hail/annotations/UnsafeRowBuilder.scala]: Nuke.

File: src/main/scala/is/hail/annotations/UnsafeRowBuilder.scala [hunk 9]
```diff
@@ -0,0 +1,359 @@
...
+          if (et == TDouble && 2500 < rightArrStart + elementOffset)
+            println(s"DEBUG double at ${ rightArrStart + elementOffset } is ${ mb.loadDouble(rightArrStart + elementOffset) }")
+        }
+        copyFromUnsafe(et, arrStart + elementOffset, rightArrStart + elementOffset, mb)
+      }
+      i += 1
+    }
+  }
+
+//  private def copyFromPointer(ft: Type, o1: Int, o2: Int, ptr: Pointer) {
```
Comment (cseed) [inline src/main/scala/is/hail/annotations/UnsafeRowBuilder.scala]: Nuke.

File: src/main/scala/is/hail/annotations/UnsafeRowBuilder.scala [hunk 10]
```diff
@@ -0,0 +1,359 @@
...
+        buffer.storeInt(start, size)
+        buffer.copyFrom(mb, start + 4, abs + 4, size)
+        if (debug)
+          println(s"copying string to $o1->$start, size=$size")
+      case st: TStruct =>
+        if (st.size > 0)
+          copyStructFromUnsafe(st, o1, o2, mb)
+      case ct: ComplexType => copyFromUnsafe(ct.representation, o1, o2, mb)
+      case coll: TContainer => copyArrayFromMem(coll.elementType, o1, o2, mb)
+      case _ => ???
```
Comment (cseed) [inline src/main/scala/is/hail/annotations/UnsafeRowBuilder.scala]: Match error seems fine here.

File: src/main/scala/is/hail/expr/AnnotationImpex.scala [hunk 11]
```diff
@@ -166,7 +166,7 @@ object SparkAnnotationImpex extends AnnotationImpex[DataType, Any] {
           Row(exportAnnotation(i.start, TLocus), exportAnnotation(i.end, TLocus))
         case TStruct(fields) =>
           if (fields.isEmpty)
-            null
+            true
```
Comment (cseed) [inline src/main/scala/is/hail/expr/AnnotationImpex.scala]: This should be `a != null`.

File: src/main/scala/is/hail/expr/AnnotationImpex.scala [hunk 12]
```diff
@@ -93,7 +93,7 @@ object SparkAnnotationImpex extends AnnotationImpex[DataType, Any] {
           Interval(importAnnotation(r.get(0), TLocus).asInstanceOf[Locus], importAnnotation(r.get(1), TLocus).asInstanceOf[Locus])
         case TStruct(fields) =>
           if (fields.isEmpty)
-            null
+            Annotation.empty
```
Comment (cseed) [inline src/main/scala/is/hail/expr/AnnotationImpex.scala]: Empty if true else null.

File: src/main/scala/is/hail/annotations/MemoryBlock.scala [hunk 13]
```diff
@@ -0,0 +1,252 @@
...
+
+  def clear() {
+    offset = 0
+  }
+
+  def result(): MemoryBlock = {
+    val reqLength = (offset + 7) / 8
+    val arr = new Array[Long](reqLength)
+    Platform.copyMemory(mb.mem, Platform.LONG_ARRAY_OFFSET, arr, Platform.LONG_ARRAY_OFFSET, offset)
+    new MemoryBlock(arr)
```
Comment (tpoterba) [inline src/main/scala/is/hail/annotations/MemoryBlock.scala]: Let's add that when there's another use case for it

File: src/main/scala/is/hail/annotations/UnsafeRow.scala (hunk 6 above)
Comment (tpoterba) [inline src/main/scala/is/hail/annotations/UnsafeRow.scala]: fair enough.

**Comment by tpoterba** [comment]
Addressed comments. Back to you.

**Pull request merged**